Frontend streamline

kestrel-x86 opened the pull request:
There are a lot of changes here, so this is going to be a long description despite my efforts to keep it compact.

### Major changes

| Unified EpisodeListItem component |
|-|
Every render of an Episode now uses the same `EpisodeListItem` component. All code that could be moved inside this component has been, which leaves it with a simpler interface, requiring fewer properties at the cost of slightly higher memory usage since the entire `Episode` struct is required.

| Unified Episode type |
|-|
When querying the database, all json responses are deserialized into the same `Episode` struct, which is a superset of all fields that were part of more specific `{name}Episode` structs. This eliminates the need for the `EpisodeTrait` and boxing of different episode types.

| Unified VirtualList type |
|-|
Rather than each page repeating code for a `VirtualList`, they can now all use the single `VirtualList` component. As with the `EpisodeListItem` changes, any code that could be moved inside this component has been and the component properties is simpler. A notable change here is that in order to enable dragging and dropping (which at this time is only on the Queue page), a `DragCallbacks` must be constructed and passed to the component, but this merely contains the same callbacks that were passed individually before this change.

| Easier downloaded episode tracking in AppState |
|-|
Replaced several fields in AppState with a new `DownloadedEpisodeRecords` struct. This struct has several methods that make adding and removing both server and local (for Tauri builds) downloads easier and less error-prone because managing episode ids and episode metadata is done automatically when adding or removing either.

### Important but less critical changes

* Reduce usage of `Option<T>` when a `None` value of that option is always treated the same. For example, a `None::<bool>` may always default to `false` so the `None` value serves no purpose. Or replacing a `Option<Vec<T>>` with an empty `Vec<T>` where checking for `None` can be replaced with checking the Vec's length.

* When setting the context menu type, use an Enum rather than strings. An enum will be checked by the compiler, while a string can have a typo and not be noticed until it causes problems.

* Separate pages and components in the file system. It just keeps things a bit more organized and easier to read when looking through `use` statements.

* Fix some `EpisodeListItem` context menus. Making sure that it always asks to *add* or *remove* a queued/saved/downloaded episode when appropriate.

* In AppState, `saved_episodes` and `saved_episode_ids` were two separate fields that would need to be updated simultaneously. To make this less error prone (in case an update to either is forgotten, `saved_episode_ids()` is now a func that returns and iterator over the ids in `saved_episodes`.

* When searching for a podcast a test was done to the endpoint to see if it connects before sending the search request. I removed this so the only request is the search. If it fails we'll know and don't need another request to determine this.

### Unimportant changes, but changes nonetheless

* Correct some indentation in `html!{}` macros because rustfmt can't.
* Time format funcs accept i32 instead of f64 because the database only stores i32 anyway.
* Loading animation is now a component `<Loading />`
* Reduce arguments that are borrows of types implementing Copy. For example, passing a `&i32` requires passing 8 bytes on a 64bit binary where a `i32` is only 4 bytes and doesn't need to be dereferenced. This will save several nanoseconds over years of heavy usage.


There are other little odds and ends that were changed and aren't worth attempting to enumerate here.

Testing UI interactions can be tricky, but I have tested this as thoroughly as possible without completely losing my mind. Notably, some parts of `tauri_downloads` can't be thoroughly tested without building that separately, but the changes there were minimal and shouldn't have changed any functionality.

kestrel-x86 committed:
Fix some localization string usage

kestrel-x86 committed:
Consolidate to Episode and EpisodeListItem

When getting episode information from the database/api, it will always be deserialized into a pod_req::Episode struct. The Episode struct contains all possible fields so it is compatible. This eliminates EpisodeTrait and the need to Box various types of Episodes.

Certain fields in the Episode have been changed from Option<T> to simply T. For example, with listen_duration a None is effectively the same value as 0. This eliminated needless unwrapping when None and the default value have no useful distinction.

This also uses EpisodeListItem to render episodes in the web ui. This still has a few bugs to work out at the moment, but the majority is working as expected. This puts all logic required to render an EpisodeListItem in episodelistitem.rs to eliminate redundant code.

Various little things throughout have been tidied up as well.

kestrel-x86 committed:
Minor cleanup

kestrel-x86 committed:
Combine all VirtualLists into virtual_list::VirtualList

kestrel-x86 committed:
I need to make smaller commits

But not today.

Clean up on_play_pause() calls. Just pass a clone of the episode instead of a handfull of fields. This also makes the function logic a little cleaner.

Change time format funcs to take i32 because the db stores listen duration as a i32 anyway.

Change podcastid to i32. I'm not entirely sure why I did this, but there was a good reason. 2 billion ids should be more than enough.

Replace more {Page}Episode types with Episode. This should take care of all of them.

Fix deserializing Episodeartwork to correct field

Rename VEpisode to EpisodeListItem

Clean up some commented-out code and superfluous comments

Consolidate some {Page}VirtualList structs into a single VirtualList. This reduces some redundant args, but requires a new struct DragCallbacks to enable dragging on a VirtualList.

kestrel-x86 committed:
Move Episode to its own file

kestrel-x86 committed:
Clean up args for on_shownotes_click

kestrel-x86 committed:
Cleaning up unused code

kestrel-x86 committed:
Use enum for ctx button page type

Eliminates problems due to typos

And cleaning up more unused code

kestrel-x86 committed:
Generate AppState.saved_episode_ids on access

Instead of storing saved_episodes and saved_episode_ids, which requires updating both simultaneously to retain parity, saved_episode_ids is now a function that generates a vec of ids from saved_episodes when it is called.

kestrel-x86 committed:
Remove unneeded accessors in Episode impl

kestrel-x86 committed:
Fix Remove From Queue menu action in Queue page

kestrel-x86 committed:
Separate pages and components

Pages contains everything directly linked to in a Route. Components contains anything else that generates HTML fragments.

Split context_button into its own file.

kestrel-x86 committed:
Rename ContextButton to ContextMenuButton

kestrel-x86 committed:
Change saved_episode_ids() to return an Iterator

kestrel-x86 committed:
Separate NotFound page and add InternalError page

kestrel-x86 committed:
Add reusable Loading component

kestrel-x86 committed:
Add optional message param to internal error page

kestrel-x86 committed:
Add DownloadedEpisodeRecords to AppState

Provides methods to make it easier and less error-prone when adding or removing a downloaded episode, wether local or on the server

kestrel-x86 committed:
Add podcast_id to create_on_title_click() args and ClickedFeedURL field

kestrel-x86 committed:
Remove connection test when searching

If it is going to fail just let it fail on the actual search request

kestrel-x86 committed:
Preparation for merge

kestrel-x86 committed:
Merge branch 'main' into frontend-streamline

kestrel-x86 committed:
Post merge cleanup

This merge was a doozy

madeofpendletonwool commented:
Hey @kestrel-x86,

Thanks again for all the effort! As mentioned earlier this is looking really great! I've started doing some digging in this evening and have noticed a couple things thus far. First, I noticed the /episode page seems to just not work anymore. So when clicking on an episode title on any of the pages. Say the feed page for example, it seems to now just land me on a empty page that doesn't load the episode content. I was able to pretty quickly get this back to working by adding the EpisodeMetadataResponse struct back in 

```
diff --git a/web/src/components/context.rs b/web/src/components/context.rs
index af19c98a..1088b46a 100644
--- a/web/src/components/context.rs
+++ b/web/src/components/context.rs
@@ -262,8 +262,8 @@ impl DownloadedEpisodeRecords {
 
     /// Remove all records of episodes downloaded on the server
     pub fn clear_server(&mut self) {
-        for id in self.local_ids.drain() {
-            if !self.server_ids.contains(&id) {
+        for id in self.server_ids.drain() {
+            if !self.local_ids.contains(&id) {
                 self.episodes.retain(|ep| ep.episodeid != id);
             }
         }
diff --git a/web/src/requests/pod_req.rs b/web/src/requests/pod_req.rs
index 8d104e47..c8221d07 100644
--- a/web/src/requests/pod_req.rs
+++ b/web/src/requests/pod_req.rs
@@ -1259,6 +1259,11 @@ pub struct EpisodeRequest {
     pub is_youtube: bool,
 }
 
+#[derive(Debug, Deserialize, Clone, PartialEq)]
+pub struct EpisodeMetadataResponse {
+    pub episode: Episode,
+}
+
 #[allow(dead_code)]
 pub async fn call_get_episode_metadata(
     server_name: &str,
@@ -1290,9 +1295,9 @@ pub async fn call_get_episode_metadata(
 
     let response_text = response.text().await?;
 
-    let ep: Episode = serde_json::from_str(&response_text)
+    let response_data: EpisodeMetadataResponse = serde_json::from_str(&response_text)
         .map_err(|e| anyhow::Error::msg(format!("Deserialization Error: {}", e)))?;
-    Ok(ep)
+    Ok(response_data.episode)
 }
 
 #[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
```

Next, I notice a pretty huge performance issue with Downloads. When you click "Download to Server" on any episode using the context menu, the entire UI becomes unresponsive - scrolling freezes, clicks don't work, everything crawls to a halt.

I believe taking a look what is probably happening is that the new unified EpisodeListItem is subscribing to AppState in yewdux. Which isn't inherently an issue, but it becomes an issue when a websocket download progress message arrives it updates app_state.downloaded_episodes. This means it's triggering every episode on the page to re-render. 

The old code used functional components that received data as props from the parent. The parent would then update only the affected episode. The new architecture moved all state access into each individual episode component, so they all react to global state changes. I'm CERTAINLY not saying the old code was better mind you. It was quite messy, but I mention it to give examples of the differences. I think we'll likely need to take at least some kind of approach that selectively updates. Perhaps a use_selector() approach or something similar. We can likely pass a few less download messages anyway, because the 2-3 per second currently is a bit much. Regardless, a selective approach will still likely be needed. 

I'll keep digging in though, otherwise looking pretty good thus far!

kestrel-x86 commented:
I think Episode is the only page I neglected to test because I didn't change much there and kind of forgot about it. I'll take a good look at it just so I can get a better understanding of how the code flows. It looks like the tricky part is getting the episode metadata from different sources if it doesn't exist in the db. 

I'll make this a priority and get back to you soon.

madeofpendletonwool commented:
Sounds good @kestrel-x86, thanks for taking a look. Feel free to let me know what other questions you might have. That differing flow indeed can be a bit tricky.

kestrel-x86 committed:
Copy Episode in AudioPlayerProps

So more fields can be accessed. In the future the redundant args should be removed.

kestrel-x86 committed:
Manually add episode.feedurl when parsing metadata

The rss feed doesn't link to itself, of course, so this needs to be done explicitly

And it turns out I was right, there is an easier way to go from a json Value to a struct.

kestrel-x86 committed:
Fix Episode page

Episode page can take a url param `episode_id` that points to an episode in the db to pull metadata from. If this is not present, the method using other params (podcast_title, etc) are used instead. The logic for getting the metadata is mostly unchanged. The global AppState is used less, and Episode pages (with appropriate url params) can be visited directly through the url.

Page state is managed in an enum that allows for Loading, Error, and Ok states.

Some field name discrepencies in Episode are corrected with serde aliases.

madeofpendletonwool committed:
Fix download performance issue causing UI freeze

madeofpendletonwool committed:
Merge remote-tracking branch 'origin/main' into frontend-streamline

kestrel-x86 commented:
I haven't really looked at downloads, and it slipped my mind until this last push from you, but its probably a good time to have an opinion now that you've fixed it :) 

Instead of sending a ui update at every N% it might be more predictable to send a task progress update at certain time frame, like every second. 

Not a priority now if your changes made it stable enough, but its something to keep in mind if his crops up again.

madeofpendletonwool committed:
Fixed issue causing test failure

madeofpendletonwool commented:
@kestrel-x86 Excellent point. It probably could be be a time based thing reasonably. I will make an issue for that. I'm going to get this merged in so that I can start making some frontend fixes but that absolutely makes more sense for a feature adjustment.

kestrel-x86 commented:
I promise not to dump giant PRs on you in the future. I got a little carried away with this one. Thanks for the merge.
